src/sdf/tape.h

```diff
@@ -0,0 +1,284 @@
+// Copyright 2024 The Manifold Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#pragma once
+
+#include <cmath>
+#include <cstdint>
+#include <cstdio>
+#include <cstring>
+#include <limits>
+
+#include "interval.h"
+#include "manifold/vec_view.h"
+
+namespace manifold::sdf {
+
+enum class OpCode : uint8_t {
+  NOP,
+  RETURN,
+  // CONST,
+
+  // unary operations
+  ABS,
+  NEG,
+  EXP,
+  LOG,
+  SQRT,
+  FLOOR,
+  CEIL,
+  ROUND,
+  SIN,
+  COS,
+  TAN,
+  ASIN,
+  ACOS,
+  ATAN,
+
+  // binary operations,
+  DIV,
+  MOD,
+  MIN,
+  MAX,
+  EQ,
+  GT,
+  AND,
+  OR,
+
+  // fast binary operations
+  ADD,
+  SUB,
+  MUL,
+  // ternary operations
+  FMA,
+  CHOICE,
+};
+
+template <typename Domain>
+struct EvalContext {
+  VecView<const uint8_t> tape;
+  VecView<Domain> buffer;
+
+  Domain& operand(uint32_t x) { return buffer[x]; }
+
+  static Domain handle_unary(OpCode op, Domain operand);
+
+  static Domain handle_binary(OpCode op, Domain lhs, Domain rhs);
+
+  static Domain handle_choice(Domain cond, Domain lhs, Domain rhs);
+
+  Domain eval() {
+    size_t i = 0;
+    while (1) {
+      OpCode current = static_cast<OpCode>(tape[i]);
+      // fast binary/ternary operations
+      if (current >= OpCode::ADD) {
```

Comment:
This is the bytecode interpreter implementation, which is optimized a bit for faster common arithmetic operations, but can still extend to other domains. It uses operator overloading for common arithmetic operations, but fallback to specialization for other functions (because I can't add a `sin` method for double, for example).